### ISSUE TITLE
Input verification

### ISSUE THREAD

**Issue by sgpyc** [gunrock/gunrock#91]
One more thing about input: if passing invalid arguments, like -undirected (missing one -), --undirect (wrong argument), --undirected --undirected (twice), --device=0, --device=1,2 (twice with different values), the behavior now is not as expected.

**Comment by wetliu** [comment]
I just wonder what are the expected results? I may try to fix them with the bogus input file issue. For example, if we passes the twice --undirected, what are the expected results?

**Comment by jowens** [comment]
I would personally replace what's in test_utils with getopt and then do whatever getopt does as its default. Or, just look at getopt's defaults. Failing any other guidance (@sgpyc?), I suggest:
- If you don't know what to do because it's not a valid flag (like `-undirected` or `-undirect`), print the error to stderr and `exit(1)`
- If you set a flag (`--undirected`) more than once, that's OK
- If you set a value more than once (`--device=0 --device=1,2`), use the rightmost (last) setting
Generally, the latter two will happen automatically if you're parsing the command line from left to right anyway.

**Comment by sgpyc** [comment]
Anything that will help to figure out what's wrong (or not so good) with the inputs will be good.
For passing the same parameter more than once (now the program will crash if using market graph as input and passing the same parameter more than once),  maybe give some warning (without quitting the program)?

Also the parameters are supposed to be typed, (e.g. non-negetive int(s) to --device, nothing or boolean to --undirected), but I think the checking now is none or hidden somewhere, not giving useful info on what is going wrong. Better to give out warning when wrong type happens (--alpha=haha, or --device=0:1).

Basically, if anything strange given in input, give a warning; if its bad enough (e.g. input file not there), give a exception and stop the program: in any way, the program should help users to make the input correct.

**Comment by wetliu** [comment]
Thank you. If you have time, can you please let me know if gunrock has the function that can stop executing the main function? If there isn't, then we can figure out the best way to solve the bad enough problems together.

**Comment by yzhwang** [comment]
Like I commented in the previous email. For detecting bogus file name. You can add something in https://github.com/gunrock/gunrock/blob/master/gunrock/app/enactor_base.cuh#L649
And exit/return error info from there.
For other command line argument handling functions, I think you should handle it here:
https://github.com/gunrock/gunrock/blob/master/gunrock/app/enactor_base.cuh#L241

Or inside CommandLineArgs class:
https://github.com/gunrock/gunrock/blob/master/gunrock/util/test_utils.h#L58

**Comment by wetliu** [comment]
Thank you! I think the comments are really helpful!

**Comment by jowens** [comment]
JDO still recommends getopt to replace CommandLineArgs. getopt has known and widely used semantics.

**Comment by yzhwang** [comment]
@jowens Thanks for the recommendation. Never heard of getopt. Just checked and it seems to be a better way. So, @wetliu wanna take the responsibility to upgrade our command line argument handling module and try using getopt? Your effort will be appreciated!

**Comment by wetliu** [comment]
I can try it. But I think I need to get more familiar with the structure of the programs first. :-)

**Comment by jowens** [comment]
@wetliu how can we most easily help you get started? The nice part is that this effort is largely self-contained, replacing CommandLineArgs with something better. I don't want to see you stuck. We want you to move forward on this.

**Comment by wetliu** [comment]
The gunrock team has provided enough support already. The team members are really helpful as they have posted  a lot of comments. Right now I am trying to fix the problem #69; this is a good start for me. If I have questions, I will comment on the github and send you emails. Thank you @jowens

**Comment by wetliu** [comment]
After talking with Yangzihao, we come up an idea of improving the input command line. We will use getopt to get the keys and corresponding values and analyze them by using two different ways. 

The first way is about the input file type. We can add one class called input file type only for market, rgg, etc, which have multiple values, such as filenames (market) or some general parameters for generating the random matrix (rgg). The class checks the values to see if they are valid under that key. 

For example, if the file type is market (key = market), the filenames will be checked if they are exist. For rgg, it will make sure that the values are numbers (not letters). 

The sample command line is:
./a.out -market=[value1,value2,...]    //no space between the filenames and values must be in brackets. 
./a.out -rgg=[value1,value2,...]       //no space between the filenames and values must be in brackets. 



The second way about the general command line arguments, which mostly have only one value for one key. Thus we can keep the map hash-table as before. 

If the users need to use more command line parameters, they can use "-" to input more command line arguments (must be seperated by space):
./a.out -rgg=[parameter1,parameter2,...] -v -s -a


Another interesting feature of getopt is that users can type short input keys, as long as the keys are not confusing with the others:   
./a.out -market=[value1,value2,...]  can change to  ./a.out -m=[value1,value2,...], as long as there exists only one word with prefix m.  

Any advice and comments will be valuable. Thank you!

**Comment by wetliu** [comment]
In terms of the input parameters, there is a question here. For example, "max-iter" uses the dash, while "quiet_mode" uses the underscore. Does that matter if we change them into the same format by using only dash or underscore?

Another question is some configuration has a flag that is different from the name in info struct, like this
info["idempotent"] =  args.CheckCmdLineFlag("idempotence");   
In this case, I wonder if it will be better to change the flag "idempotence" to "idempotent" to make the name and the flag the same?

**Comment by jowens** [comment]
I encourage using the underscore rather than the dash, since with an underscore it's a valid variable name. At any rate, make it consistent. I think you should also make all flags/stats entries the same as well. I am agnostic as to which name to choose (others may have a preference).

**Comment by wetliu** [comment]
Cool, thanks.

**Comment by sgpyc** [comment]
The command line parameters are handled by util::Parameters in the refactored code.
https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/util/parameters.h#L107
getopt is part of it.
https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/util/parameters.h#L468

The validation of command line parameters are in effect:
```
sgpyc@mario:~/Projects/gunrock/tests/sssp$ ./bin/test_sssp_9.1_x86_64 rmat --graph_scale=10 --graph-scale=12 --graph-scale=13 -graph-scale
./bin/test_sssp_9.1_x86_64: unrecognized option '--graph_scale=10'
Warnning : Parameter graph-scale(../../gunrock/graphio/graphio.cuh:105) specified more than once, only latter value 13 is effective.
./bin/test_sssp_9.1_x86_64: option '--graph-scale' requires an argument
...
```
